crates/sui-graphql-rpc/schema/draft_schema.graphql

```diff
@@ -191,6 +191,58 @@ scalar DateTime
 #   | { Struct:  [{ name: string, value: MoveData }] }
 scalar MoveData
 
+# The signature of a concrete Move Type (a type with all its type
+# parameters instantiated with concrete types, that contains no
+# references), corresponding to the following recursive type:
+#
+# type MoveTypeSignature =
+#     "address"
+#   | "bool"
+#   | "u8" | "u16" | ... | "u256"
+#   | { vector: MoveTypeSignature }
+#   | {
+#       package: string,
+#       module: string,
+#       type: string,
+#       typeParameters: [MoveTypeSignature]?,
+#     }
+scalar MoveTypeSignature
+
+# The shape of a concrete Move Type (a type with all its type
+# parameters instantiated with concrete types), corresponding to the
+# following recursive type:
+#
+# type MoveTypeLayout =
+#     "address"
+#   | "bool"
+#   | "u8" | "u16" | ... | "u256"
+#   | { vector: MoveTypeLayout }
+#   | { struct: [{ name: string, layout: MoveTypeLayout }] }
+scalar MoveTypeLayout
+
+# The shape of an abstract Move Type (a type that can contain free
+# type parameters, and can optionally be taken by reference),
+# corresponding to the following recursive type:
+#
+# type OpenMoveTypeSignature = {
+#   ref: ("&" | "&mut")?,
+#   body: OpenMoveTypeSignatureBody,
+# }
+#
+# type OpenMoveTypeSignatureBody =
+#     "address"
+#   | "bool"
+#   | "u8" | "u16" | ... | "u256"
+#   | { vector: OpenMoveTypeSignatureBody }
+#   | {
+#       package: string,
+#       module: string,
+#       type: string,
+#       typeParameters: [OpenMoveTypeSignatureBody]?
+#     }
+#   | { TypeParameter: number }
+scalar OpenMoveTypeSignature
```

Comment:
One thing that I think could be helpful here would be to have a version of this where references to other types are resolved.

GraphQL doesn't have a way to take the results of one query and pipe them as arguments of a nested call, so to fully resolve a nested type would require multiple GraphQL queries.


Getting back something like:
```js
{
  vector: {
    package: 'a',
    module: 'b',
    type: 'c',
    typeParameters: ['u8', { TypeParameter: 1 }],
 }
}
```

Isn't sufficient for actually building out a BCS schema, having a way to get a generic layout would be really helpful here:

```js
{
  vector: {
    package: 'a',
    module: 'b',
    type: 'c',
    typeParameters: ['u8', { TypeParameter: 1 }],
    layout: {
      struct: [{
        name: 'id',
        layout: 'u8', // typeParamaters[0]
     }, {
        name: 'inner',
        layout: { TypeParameter: 1 } // typeParameters[2]
     }
  }
}
```

this would allow us to create the corresponding BCS schema:

```ts
const type = bcs.generic(['T'], (T) => bcs.vector(
  bcs.struct({
    id: bcs.u8(),
    inner: T,
  })
))
```

Reply:
I was thinking about this use case, and it sounds like `MoveStruct` might be the ideal thing to query for this? Here's what the relevant part of the schema:

```graphql
type MoveStruct {
  id : ID!
  module: MoveModule!
  name: String!
  abilities: [MoveAbility]
  typeParameters: [MoveStructTypeParameter]
  fields: [MoveField]
}

type MoveStructTypeParameter {
  constraints: [MoveAbility]
  isPhantom: Boolean
}

enum MoveAbility {
  COPY
  DROP
  STORE
  KEY
}

type MoveField {
  name: String
  type: OpenMoveType
}

type OpenMoveType {
  repr: String
  signature: OpenMoveTypeSignature
}
```

Let me know what you think, but I believe a query of the `MoveStruct` corresponding to `type` in your example would give you enough information to generate its BCS representation (i.e. its use of `OpenMoveTypeSignature` exposes just enough information about field types to write the body of the serializer).